book/src/02_development/05_cheatcodes.md

```diff
@@ -0,0 +1,156 @@
+
+- coverage
+- tracing
+- debugger
+- logging
+
+## [Cheatcode inspector](../../crates/cheatcodes/src/inspector.rs)
+
+The concept of cheatcodes and cheatcode inspector is very simple.
+
+Cheatcodes are calls to a specific address, the cheatcode handler address, defined as `address(uint160(uint256(keccak256("hevm cheat code"))))` (`0x7109709ECfa91a80626fF3989D68f67F5b1DD12D`).
+
+In Solidity, this can be initialized as `Vm constant vm = Vm(0x7109709ECfa91a80626fF3989D68f67F5b1DD12D);`, but generally this is inherited from `forge-std/Test.sol`.
+
+Since cheatcodes are bound to a constant address, the cheatcode inspector listens for that address:
+
+```rust
+impl Inspector for Cheatcodes {
+    fn call(
+        &mut self,
+        data: &mut EVMData<'_, DB>,
+        call: &mut CallInputs,
+        is_static: bool,
+    ) -> (Return, Gas, Bytes) {
+        if call.contract == CHEATCODE_ADDRESS {
+            // intercepted cheatcode call
+            // --snip--
+        }
+    }
+}
+```
+
+When a call to a cheatcode is intercepted we try to decode the calldata into a known cheatcode.
+
+Rust bindings for the cheatcode interface are generated via the [Alloy](https://github.com/alloy-rs) [`sol!`](https://docs.rs/alloy-sol-macro/latest/alloy_sol_macro/macro.sol.html) macro.
+
+If a call was successfully decoded into the `VmCalls` enum that the `sol!` macro generates, the last step is a large `match` over the decoded function call structs, which serves as the implementation handler for the cheatcode. This is also automatically generated, in part, by the `sol!` macro, through the use of a custom internal derive procedural macro.
+
+## Cheatcodes implementation
+
+All the cheatcodes are defined in a large [`sol!`] macro call in [`cheatcodes/spec/src/vm.rs`]:
+
+```rust
+sol! {
+#[derive(Cheatcode)]
+interface Vm {
+    //  ======== Types ========
+
+    /// Error thrown by a cheatcode.
+    error CheatcodeError(string message);
+
+    // ...
+
+    // ======== EVM ========
+
+    /// Gets the address for a given private key.
+    #[cheatcode(group = Evm, safety = Safe)]
+    function addr(uint256 privateKey) external pure returns (address keyAddr);
+
+    /// Gets the nonce of an account.
+    #[cheatcode(group = Evm, safety = Safe)]
+    function getNonce(address account) external view returns (uint64 nonce);
+
+    // ...
+}
+}
+```
+
+This, combined with the use of an internal [`Cheatcode` derive macro](#cheatcode-derive-macro), allows us to generate both the Rust definitions and the JSON specification of the cheatcodes.
+
+Cheatcodes are manually implemented through the [`Cheatcode` trait](#cheatcode-trait), which is called in the [`Cheatcodes` inspector](#cheatcode-inspector) implementation.
+
+### [`sol!`]
+
+Generates the raw Rust bindings for the cheatcodes, as well as lets us specify custom attributes individually for each item, such as functions and structs, or for entire interfaces.
```

Comment:
```suggestion
Generates the raw Rust bindings for the cheatcodes, as well as letting us specify custom attributes individually for each item, such as functions and structs, or for entire interfaces.
```

Reply:
Right, lost after rebase, sorry about that